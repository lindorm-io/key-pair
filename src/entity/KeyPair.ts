import Joi from "joi";
import { Algorithm, KeyPairEvent, KeyType, NamedCurve } from "../enum";
import { IJoseData, IJwk, IKeyJwk } from "../types";
import { JOI_KEY_ALGORITHM, JOI_KEY_ALGORITHMS, JOI_KEY_NAMED_CURVE, JOI_KEY_TYPE } from "../constant";
import { decodeKeys, encodeKeys } from "../util";
import { includes, orderBy } from "lodash";
import {
  EntityBase,
  EntityCreationError,
  IEntity,
  IEntityAttributes,
  IEntityOptions,
  JOI_ENTITY_BASE,
} from "@lindorm-io/entity";

export interface IKeyPairAttributes extends IEntityAttributes {
  algorithms: Array<Algorithm>;
  allowed: boolean;
  expires: Date | null;
  namedCurve: NamedCurve | null;
  passphrase: string | null;
  preferredAlgorithm: Algorithm;
  privateKey: string | null;
  publicKey: string;
  type: KeyType;
}

export interface IKeyPairOptions extends IEntityOptions {
  algorithms: Array<Algorithm>;
  allowed?: boolean;
  expires?: Date;
  namedCurve?: NamedCurve;
  passphrase?: string;
  preferredAlgorithm?: Algorithm;
  privateKey?: string;
  publicKey: string;
  type: KeyType;
}

const schema = Joi.object({
  ...JOI_ENTITY_BASE,

  algorithms: JOI_KEY_ALGORITHMS.required(),
  allowed: Joi.boolean().required(),
  expires: Joi.date().allow(null).required(),
  namedCurve: JOI_KEY_NAMED_CURVE.allow(null).required(),
  passphrase: Joi.string().allow(null).required(),
  preferredAlgorithm: JOI_KEY_ALGORITHM.required(),
  privateKey: Joi.string().allow(null).required(),
  publicKey: Joi.string().required(),
  type: JOI_KEY_TYPE.required(),
});

export class KeyPair extends EntityBase<IKeyPairAttributes> implements IEntity<IKeyPairAttributes> {
  public readonly algorithms: Array<Algorithm>;
  public readonly namedCurve: NamedCurve | null;
  public readonly passphrase: string | null;
  public readonly privateKey: string | null;
  public readonly publicKey: string;
  public readonly type: KeyType;
  private _allowed: boolean;
  private _expires: Date | null;
  private _preferredAlgorithm: Algorithm;

  public constructor(options: IKeyPairOptions) {
    super(options);

    this._allowed = options.allowed !== false;
    this._expires = options.expires || null;
    this._preferredAlgorithm = options.preferredAlgorithm || orderBy(options.algorithms, [(item) => item], ["desc"])[0];

    this.algorithms = options.algorithms;
    this.namedCurve = options.namedCurve || null;
    this.passphrase = options.passphrase || null;
    this.privateKey = options.privateKey || null;
    this.publicKey = options.publicKey;
    this.type = options.type;
  }

  public get allowed(): boolean {
    return this._allowed;
  }
  public set allowed(allowed: boolean) {
    this._allowed = allowed;
    this.addEvent(KeyPairEvent.ALLOWED_CHANGED, { allowed: this._allowed });
  }

  public get expires(): Date | null {
    return this._expires;
  }
  public set expires(expires: Date | null) {
    this._expires = expires;
    this.addEvent(KeyPairEvent.EXPIRES_CHANGED, { expires: this._expires });
  }

  public get preferredAlgorithm(): Algorithm {
    return this._preferredAlgorithm;
  }
  public set preferredAlgorithm(preferredAlgorithm: Algorithm) {
    if (!includes(this.algorithms, preferredAlgorithm)) {
      throw new Error("Invalid preferredAlgorithm");
    }

    this._preferredAlgorithm = preferredAlgorithm;
    this.addEvent(KeyPairEvent.PREFERRED_ALGORITHM_CHANGED, { preferredAlgorithm: this._preferredAlgorithm });
  }

  public create(): void {
    for (const evt of this.events) {
      if (evt.name !== KeyPairEvent.CREATED) continue;
      throw new EntityCreationError(this.constructor.name);
    }

    this.addEvent(KeyPairEvent.CREATED, this.toJSON());
  }

  public getKey(): string {
    return this.id;
  }

  public async schemaValidation(): Promise<void> {
    await schema.validateAsync(this.toJSON());
  }

  public toJSON(): IKeyPairAttributes {
    return {
      ...this.defaultJSON(),

      algorithms: this.algorithms,
      allowed: this.allowed,
      expires: this.expires,
      namedCurve: this.namedCurve,
      passphrase: this.passphrase,
      preferredAlgorithm: this.preferredAlgorithm,
      privateKey: this.privateKey,
      publicKey: this.publicKey,
      type: this.type,
    };
  }

  public toJWK(exposePrivateKey = false): IJwk {
    const data: IKeyJwk = encodeKeys({
      exposePrivateKey,
      namedCurve: this.namedCurve,
      privateKey: this.privateKey,
      publicKey: this.publicKey,
      type: this.type,
    });

    return {
      alg: this.preferredAlgorithm,
      created: this.created.getTime(),
      crv: this.namedCurve ? this.namedCurve : undefined,
      expires: this.expires ? this.expires.getTime() : undefined,
      key_ops: ["sign", "verify"],
      kid: this.id,
      kty: this.type,
      use: "sig",
      ...data,
    };
  }

  public static fromJWK(jwk: IJwk): KeyPair {
    const data: IJoseData = decodeKeys(jwk);

    return new KeyPair({
      id: jwk.kid,
      algorithms: [jwk.alg as Algorithm],
      created: jwk.created ? new Date(jwk.created) : undefined,
      expires: jwk.expires ? new Date(jwk.expires) : undefined,
      namedCurve: jwk.crv ? (jwk.crv as NamedCurve) : undefined,
      preferredAlgorithm: jwk.alg as Algorithm,
      type: jwk.kty as KeyType,
      ...data,
    });
  }
}
