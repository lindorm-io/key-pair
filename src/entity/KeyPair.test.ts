import MockDate from "mockdate";
import { KeyPair } from "./KeyPair";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "mock-uuid"),
}));

MockDate.set("2020-01-01 08:00:00.000");
const date = new Date("2020-01-01 08:00:00.000");

describe("KeyPair.ts", () => {
  let keyPair: KeyPair;

  beforeEach(() => {
    keyPair = new KeyPair({
      algorithm: "algorithm",
      expires: date,
      passphrase: "passphrase",
      privateKey: "privateKey",
      publicKey: "publicKey",
      type: "type",
    });
  });

  test("should have all data", () => {
    expect(keyPair).toMatchSnapshot();
  });

  test("should have optional data", () => {
    keyPair = new KeyPair({
      algorithm: "algorithm",
      type: "type",
    });

    expect(keyPair).toMatchSnapshot();
  });

  test("should create", () => {
    keyPair.create();
    expect(keyPair.events).toMatchSnapshot();
  });

  test("should get algorithm", () => {
    expect(keyPair.algorithm).toBe("algorithm");
  });

  test("should get/set expires", () => {
    expect(keyPair.expires).toBe(date);

    const expires = new Date("2021-01-01 00:00:01");
    keyPair.expires = expires;

    expect(keyPair.expires).toBe(expires);
    expect(keyPair.events).toMatchSnapshot();
  });

  test("should get passphrase", () => {
    expect(keyPair.passphrase).toBe("passphrase");
  });

  test("should get privateKey", () => {
    expect(keyPair.privateKey).toBe("privateKey");
  });

  test("should get publicKey", () => {
    expect(keyPair.publicKey).toBe("publicKey");
  });

  test("should get type", () => {
    expect(keyPair.type).toBe("type");
  });
});
