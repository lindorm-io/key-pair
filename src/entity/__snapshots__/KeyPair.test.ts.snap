// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeyPair.ts should create 1`] = `
Array [
  Object {
    "date": 2020-01-01T07:00:00.000Z,
    "name": "key_pair_created",
    "payload": Object {
      "algorithm": "algorithm",
      "expires": 2020-01-01T07:00:00.000Z,
      "passphrase": "passphrase",
      "privateKey": "privateKey",
      "publicKey": "publicKey",
      "type": "type",
    },
  },
]
`;

exports[`KeyPair.ts should get/set expires 1`] = `
Array [
  Object {
    "date": 2020-01-01T07:00:00.000Z,
    "name": "key_pair_expires_changed",
    "payload": Object {
      "expires": 2020-12-31T23:00:01.000Z,
    },
  },
]
`;

exports[`KeyPair.ts should have all data 1`] = `
KeyPair {
  "_algorithm": "algorithm",
  "_created": 2020-01-01T07:00:00.000Z,
  "_events": Array [],
  "_expires": 2020-01-01T07:00:00.000Z,
  "_id": "mock-uuid",
  "_passphrase": "passphrase",
  "_privateKey": "privateKey",
  "_publicKey": "publicKey",
  "_type": "type",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;

exports[`KeyPair.ts should have optional data 1`] = `
KeyPair {
  "_algorithm": "algorithm",
  "_created": 2020-01-01T07:00:00.000Z,
  "_events": Array [],
  "_expires": null,
  "_id": "mock-uuid",
  "_passphrase": null,
  "_privateKey": null,
  "_publicKey": null,
  "_type": "type",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;
